src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Callables.cs

```diff
@@ -207,9 +207,26 @@ internal sealed partial class ExpressionEval {
             return instance.Call(p.Name, ArgumentSet.Empty);
         }
 
-        private IMember TryEvaluateWithArguments(IPythonModule module, FunctionDefinition fd, IArgumentSet args) {
+
+        private readonly Dictionary<int, IMember> _argEvalCache = new Dictionary<int, IMember>();
+
+        private IMember TryEvaluateWithArguments(IPythonFunctionType fn, IArgumentSet args) {
+            var name = fn.DeclaringType != null ? $"{fn.DeclaringModule.Name}.{fn.Name}" : fn.Name;
+            var argHash = args
+                .Arguments
+                .Select(a => a.Name.GetHashCode() ^ 397 * (a.Value?.GetHashCode() ?? 0))
+                .Aggregate(0, (current, d) => 31 * current ^ d);
+            var key = fn.DeclaringModule.Name.GetHashCode() ^ name.GetHashCode() ^ (397 * argHash);
+
+            if (_argEvalCache.TryGetValue(key, out var result)) {
+                return result;
```

Comment:
Won't hash collisions here make some unrelated functions have the same evaluation?

Reply:
Functions have fully qualified names like module.class.method which should have unique hashes